File: Cuarto paradigma/Excepciones/exepciones java/src/main/java/com/mycompany/exepciones/Exepciones.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Project/Maven2/JavaApp/src/main/java/${packagePath}/${mainClassName}.java to edit this template
 */

package com.mycompany.exepciones;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import java.util.Set;

/**
 *
 * @author dinar
 */
public class Exepciones {
    static ArrayList<Object> listaObjetos = new ArrayList<Object>();
    static ArrayList<String> listaOperaciones = new ArrayList<String>();
    
    public static boolean isNumeric(String cadena){
        try{
            Integer.parseInt(cadena);
            return true;
        }catch(NumberFormatException nfe){
            return false;
        }
    }
    
    public static void mostrarListaObjetos(){
        for (int i = 0; i < listaObjetos.size(); i ++){
            if(listaObjetos.get(i) instanceof Num == true){
                System.out.println( ((Num)listaObjetos.get(i)).getNumero_var());
            }
            else{
                System.out.println( ((Op)listaObjetos.get(i)).getOperacion_var());
            }
        }
    }
    
    public static void descomponer ()throws ErrorCF{
        ArrayList<String> listaCaracteres = new ArrayList<String>();
        Scanner datos = new Scanner(System.in);
        System.out.print("Ingrese la operacion: ");
        
        String operacionNueva = datos.nextLine();
        System.out.println("");
        System.out.println(operacionNueva);
        String []listaDatos = operacionNueva.split("");
        for(String letra: listaDatos){
            listaCaracteres.add(letra);
        }
        
        //Try
        boolean validarExpresion = true;
        boolean validarCaracter = true;
        int i = 0;
        while (i < listaCaracteres.size()){
            if (i == 0){
                if (isNumeric(listaCaracteres.get(i))){
                    validarExpresion = true;
                    Num n = new Num(Integer.parseInt(listaCaracteres.get(i)));
                    listaObjetos.add(n);
                }
                else{
                    validarExpresion = false;
                    if (isNumeric(listaCaracteres.get(i)) || listaOperaciones.contains(listaCaracteres.get(i))){
                        validarCaracter = true;
                    }
                    else{
                        validarExpresion = false;
                    }
                    break;
                }
            }
            else if (i == 1){
                if (listaOperaciones.contains(listaCaracteres.get(i))){
                    validarExpresion = true;
                    Op o = new Op(listaCaracteres.get(i));
                    listaObjetos.add(o);
                }
                else{
                    validarExpresion = false;
                    if (isNumeric(listaCaracteres.get(i)) || listaOperaciones.contains(listaCaracteres.get(i))){
                        validarCaracter = true;
                    }
                    else{
                        validarExpresion = false;
                    }
                    break;
                }
            }
            else if ((i % 2) == 0){
                if (isNumeric(listaCaracteres.get(i))){
                    validarExpresion = true;
                    Num n = new Num(Integer.parseInt(listaCaracteres.get(i)));
                    listaObjetos.add(n);
                }
                else{
                    validarExpresion = false;
                    if (isNumeric(listaCaracteres.get(i)) || listaOperaciones.contains(listaCaracteres.get(i))){
                        validarCaracter = true;
                    }
                    else{
                        validarExpresion = false;
                    }
                    break;
                
                }
            }
            else{
                if (listaOperaciones.contains(listaCaracteres.get(i))){
                    validarExpresion = true;
                    Op o = new Op(listaCaracteres.get(i));
                    listaObjetos.add(o);
                    
                }
                else{
                    validarExpresion = false;
                    if (isNumeric(listaCaracteres.get(i)) || listaOperaciones.contains(listaCaracteres.get(i))){
                        validarCaracter = true;
                    }
                    else{
                        validarExpresion = false;
                    }
                    break;
                }
            }
            i = i + 1;
            
        }
        
        if (validarCaracter && validarExpresion){
        System.out.println("Lista con objetos llena!");
        mostrarListaObjetos();

        }
        if  (!validarCaracter){
            throw new ErrorCF("La expresion presenta un error de caracteres.");
        }
        if  (!validarExpresion){
            throw new ErrorCF("La expresion presenta un error de expresion.");
        }
        
        
    }
    public static void calcular(){
        int num1 = 0;
        int num2 = 0;
        String operacion = "";
        int resultado = 0;
        while (listaObjetos.size() >= 3){
            num1 = ((Num)listaObjetos.get(0)).getNumero_var();
            operacion = ((Op)listaObjetos.get(1)).getOperacion_var();
            num2 = ((Num)listaObjetos.get(2)).getNumero_var();
            
            if (operacion.equals("+")){
                resultado = num1 + num2;
            }
            if (operacion.equals("-")){
                
                resultado = num1 - num2;
            }
            if (operacion.equals("*")){
                
                resultado = num1 * num2;
            }
            if (operacion.equals("/")){
                
                resultado = num1 / num2;
            }
            Num nuevoN = new Num(resultado);
            listaObjetos.remove(0);
            listaObjetos.remove(0);
            listaObjetos.remove(0);
            listaObjetos.add(0, nuevoN);
            System.out.println("Lista nueva");
            mostrarListaObjetos();
        }
    }
    
    public static void main(String[] args) {
        listaOperaciones.add("+");
        listaOperaciones.add("*");
        listaOperaciones.add("-");
        listaOperaciones.add("/");
        try{
            descomponer();
            calcular();
        }catch(ErrorCF e){
            System.out.println("Error capturado: " + e.getMessage());
            
        }
        
        //String text= new String("Hello, My name is Sachin");  
        /* Splits the sentence by the delimeter passed as an argument */  
        //String[] sentences = text.split("");  
        //System.out.println(Arrays.toString(sentences));  
        
    }
}

File: Cuarto paradigma/Excepciones/excepciones python/exepciones.py


from cmath import inf
from mimetypes import init

class Num:
    def __init__(self, numero):
        self.n = numero

class Op:
    def __init__(self, operacion):
        self.operacion = operacion

class ErrorCaracter_Expresion(Exception):
    def __init__(self, mensaje, informacion):
        super().__init__()
        self.mensaje = mensaje
        self.informacion = informacion




listaObjetos = []
listaOperaciones = ['+','-','/','*']
def descomponer():
    listaDatos = []
    datos = input("Ingrese las operaciones y numeros en infijo: ")
    for letra in datos:
        listaDatos.append(letra)
    print(listaDatos)

    try:
        i = 0
        validarExpresion = True
        validarCaracter = True
        while i < len(listaDatos):
            if i == 0:
                if listaDatos[i].isdigit():
                    validarExpresion = True
                    num = Num(int(listaDatos[i]))
                    listaObjetos.append(num)
                else:
                    validarExpresion = False
                    if listaDatos[i].isdigit() or listaDatos[i] in listaOperaciones:
                        validarCaracter = True
                    else:
                        validarCaracter = False
                    break
            
            elif i == 1:
                if listaDatos[i] in listaOperaciones:
                    validarExpresion = True
                    op = Op(listaDatos[i])
                    listaObjetos.append(op)
                else:
                    validarExpresion = False
                    if listaDatos[i].isdigit() or listaDatos[i] in listaOperaciones:
                        validarCaracter = True
                    else:
                        validarCaracter = False
                    break
            
            elif (i % 2 == 0):
                if listaDatos[i].isdigit():
                    validarExpresion = True
                    num = Num(int(listaDatos[i]))
                    listaObjetos.append(num)
                else:
                    validarExpresion = False
                    if listaDatos[i].isdigit() or listaDatos[i] in listaOperaciones:
                        validarCaracter = True
                    else:
                        validarCaracter = False
                    break
            
            else:
                if listaDatos[i] in listaOperaciones:
                    validarExpresion = True
                    op = Op(listaDatos[i])
                    listaObjetos.append(op)
                else:
                    validarExpresion = False
                    if listaDatos[i].isdigit() or listaDatos[i] in listaOperaciones:
                        validarCaracter = True
                    else:
                        validarCaracter = False
                    break

            i = i + 1
        if validarExpresion and validarCaracter:
            print("Lista con los objetos esta lista para trabajar!")
            verListaObjetos()
        if not(validarCaracter):
            raise ErrorCaracter_Expresion("La expresion presenta un error en caracteres: ", datos)
        if not(validarExpresion):
            raise ErrorCaracter_Expresion("La expresion presenta no se cuenta en infijo: ", datos)

    except ErrorCaracter_Expresion as e:
        print(e.mensaje, e.informacion)
def ifverListaObjetos():
    for ob in listaObjetos:
        if type(ob) is Num:
            print(ob.n)
        if type(ob) is Op:
            print(ob.operacion)
def calcular():
    numero1 = 0
    numero2 = 0
    resultado = 0

    while len(listaObjetos) >= 3:
        numero1 = listaObjetos[0].n
        numero2 = listaObjetos[2].n
        resultado = 0
        operar = listaObjetos[1].operacion
        if operar == '+':
            resultado = numero1 + numero2
        if operar == '-':
            resultado = numero1 - numero2

        if operar == '*':
            resultado = numero1 * numero2

        if operar == '/':
            resultado = numero1 / numero2

        newnum = Num(resultado)
        listaObjetos.pop(0)
        listaObjetos.pop(0)
        listaObjetos.pop(0)
        listaObjetos.insert(0,newnum)

    verListaObjetos()

#5*2/4+3
descomponer()
calcular()